src/crypto/hmac_sha256.cpp

```diff
@@ -11,7 +11,9 @@ CHMAC_SHA256::CHMAC_SHA256(const unsigned char* key, size_t keylen)
     unsigned char rkey[64];
     if (keylen <= 64) {
         memcpy(rkey, key, keylen);
-        memset(rkey + keylen, 0, 64 - keylen);
+        if (keylen < 64) {
```

Comment:
I… don't think this makes the code any more understandable, I had to think for a while to understand why this was `<64` and why one'd want to skip the memset if the data is smaller than that.
Calling memset with size 0 is no problem. If calling it with a NULL pointer is, then a check for a null pointer should be added, not a check on the length.

Reply:
Oh, I might have been a bit brief in the OP - let me clarify:

The goal of this PR is _not_ try to avoid `memset(nullptr, 0, 0)` (plus: `rkey + keylen` cannot be `nullptr`).

The goal of the PR is to avoid `memset(p, 0, 0)` where `p` is a pointer which is outside of the object and thus not valid for writing.

Please note that `rkey + 64` is outside of the object.

See these discussions for context:
* https://github.com/bitcoin/bitcoin/pull/15950#issuecomment-489461791
* https://stackoverflow.com/questions/29844298/is-it-legal-to-call-memcpy-with-zero-length-on-a-pointer-just-past-the-end-of-an

Please re-review in light of this clarification :-)

Should I add a comment in the code making the rationale more clear?